gcpp/programs/cfm.1
.TH cfm 1 "August 2013" "cmassoc-tools-1.9.1" "Motley Tools"

.SH NAME
cfm - format C language source code

.SH SYNOPSIS
.BR cfm
.RI [ options ]
.IR file
.RI [ file ] 
.RI [ ... ]

.PP
.BR cfm
.RI [ options ]
.RI [< stdin ]
.RI [> stdout ]

.SH DESCRIPTION

.PP
Format C Language source code files and expand or update selected comment text.

.SH OPTIONS
Long options are not supported.
All options are single letters.

.TP
.RB - b
Indent lines using 3 spaces (blanks) instead of one tab.

.TP
.RB - c
Convert C++ comments to C style comments.

.TP
.RB - C
Format comments only.
Leave source code unformatted.

.TP
-\fBf\fI profile
The name of the profile to use.
The presence of this option may affect option \fB-g\fR.
The default profile is \fB/etc/cfm.ini\fR.

.TP
-\fBg\fI heading\fR
The profile heading that defines the expansion text to be used.
If this option is omitted or the named heading is absent then default text will be used.

.TP
.RB - h
Insert an include guard before the first processor statement in the file.
The include guard name is the source file basename in upper case suffixed by "_HEADER".

.TP
.RB - k
Prevent future modification of the preamble, license, publish and/or release comment lines by discarding the distinguishing character that identifies them.
The associated comments are updated but the distinguishing characters are not replaced.
This means that successive passes of this program cannot detect those lines will not modify them.
Use option \fB-l\fR, \fB-p\fR, \fBP\fR and/or \fB-f\fR to specify the type of comment lines to protect.

.TP
.RB - l
Expand preamble comment lines starting with "\fB *.\fR" using the \fBlicense\fR string.
If this option is present with option \fB-x\fR then license comment lines are discarded.
If this option is present with option \fB-k\fR then license comment lines are protected.
If this option is omitted then preamble license lines are not updated.

.TP
.RB - m
Convert single-line comments to multi-line comments.

.TP
.RB - o
Print a default profile on stdout.
This option can be used to create a new profile which can then be modified.

.TP
.RB - p
Expand preamble comments lines starting with "\fB *:\fR" using the \fBpublish\fR string.
If this option is present with option \fB-x\fR then publish comment lines are discarded.
If this option is present with option \fB-k\fR then publish comment lines are protected.
If this option is omitted then preamble publish lines are not updated.

.TP
.RB - P
Expand preamble comments lines starting with "\fB *~\fR" using the \fBpreamble\fR string.
If this option is present with option \fB-x\fR then publish comment lines are discarded.
If this option is present with option \fB-k\fR then publish comment lines are protected.
If this option is omitted then preamble publish lines are not updated.

.TP
.RB - r
Expand preamble comments lines starting with "\fB *;\fR" using the \fBrelease\fR string.
If this option is present with option \fB-x\fR then release comment lines are discarded.
If this option is present with option \fB-k\fR then release comment lines are protected.
If this option is omitted then release comment lines are not updated.

.TP
.RB - s
Insert an include guard before the first processor statement in the file.
The include guard name is the source file basename in upper case suffixed by "_SOURCE".

.TP
.RB - S
Format comments and source code.

.TP
.RB - t
Indent text using one tab instead of 3 spaces.

.TP
.RB - q
Quite mode.
This option has no effect at this time.

.TP
.RB - v
Verbose mode.
This option has no effect at this time.

.TP
-\Fbw\fI width\fR
The width of comment bars.
The default width is 68 columns which exactly matches the width of a vi editor line when numbering is enabled.

.TP
.RB - x
Discard the preamble, license, publish and/or release comment lines and their distinguishing characters.
One blank comment line remains in place of each discarded comment line.
Use option \fB-l\fR, \fB-p\fR, \fBP\fR and/or \fB-f\fR to specify the type of comment lines to discard.

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over other options on the command line.

.TP
.RB - ! ,-- version
Print program version information on stdout.
This option takes precedence over other options on the command line.
Usethis option when sending screen dumps to Intellon Technical Support so that they know exactly which version of the Linux Toolkit you are using.

.SH ARGUMENTS

.TP
.BR file
The name of a C Language source file.
This file is replaced with the formatted output and a numbered backup file is created.
Attempts to format syntactically incorrect source files may result in unpredictable output or program errors.

.SH FILES

.TP
/etc/cfm.ini
The default profile containing named groups of definitions for "package", "release", "license", "include" and "special" strings.
The derault profile can be printed using option \fB-o\fR and modified.

.SH EXAMPLES
The following example formats source file \fBabc.c\fR.
The backup file \fBabc.c.nnn\fR is automaticallu created in case recovery is required.

.PP
   # cfm abc.c

.PP
The following example prints the default profile on stdout.
The \fBpackage\fR string appears on comment lines starting with period.
The \fBrelease\fR string appears on comment lines starting with colon.
The \fBlicense\fR string appears on comment lines starting with semicolon.
Other strings that may be defined have special application that that shall remain undocumented.

.PP
   # cfm -o > /etc/cfm.ini
   [default]
   package=Motley Tools by Charles Maier; cmaier@cmassoc.net;
   release=Published 1982-2005 by Charles Maier for personal use;
   license=Released under GNU General Public License v2 and later;

.PP
The following example creates the default profile, \fB/etc/cfm.ini\fR.
You can then edit the file by redefining properties or adding additional sections that can be referenced by name using option \fB-g\fR.

.PP
   # cfm -o > /etc/cfm.ini

.SH SEE ALSO
.BR afm ( 7 ),
.BR ctidy ( 7 ),
.BR csstidy ( 7 ),
.BR phptidy ( 7 ),
.BR xmltidy ( 7 )

.SH AUTHORS
 Charles Maier <cmaier@cmassoc.net>



gcpp/programs/cfm.cpp
/*====================================================================*
 *
 *   cfm.cpp - format C/C++ source files;
 *
 *   custom C/C++ source code formatter;
 *
 *.  Motley Tools by Charles Maier <cmaier@cmassoc.net>;
 *:  Copyright 2001-2006 by Charles Maier Associates;
 *;  Licensed under the Internet Software Consortium License
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <cstdlib>
#include <iostream>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../classes/ogetoptv.hpp"
#include "../classes/ofileopen.hpp"
#include "../classes/oprofile.hpp"
#include "../classes/octidy.hpp"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../classes/ogetoptv.cpp"
#include "../classes/oputoptv.cpp"
#include "../classes/oversion.cpp"
#include "../classes/oerror.cpp"
#include "../classes/oflagword.cpp"
#include "../classes/ofileopen.cpp"
#include "../classes/ofilespec.cpp"
#include "../classes/okeywords.cpp"
#include "../classes/oprofile.cpp"
#include "../classes/ocontext.cpp"
#include "../classes/owildcard.cpp"
#include "../classes/oescape.cpp"
#include "../classes/oswitch.cpp"
#include "../classes/otext.cpp"
#include "../classes/oascii.cpp"
#include "../classes/oinclude.cpp"
#include "../classes/ocomment.cpp"
#include "../classes/ocgotowords.cpp"
#include "../classes/ocexitwords.cpp"
#include "../classes/oclangwords.cpp"
#include "../classes/octidy.cpp"
#include "../classes/osource.cpp"
#include "../classes/oindent.cpp"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define PROFILE_NAME "/etc/cfm.ini"     
#define SECTION_NAME "default"     

/*====================================================================*
 *   
 *   int main (int argc, char const * argv []);
 *   
 *   
 *.  Motley Tools by Charles Maier <cmaier@cmassoc.net>;
 *:  Copyright 2001-2006 by Charles Maier Associates;
 *;  Licensed under the Internet Software Consortium License
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv []) 

{ 
	static char const * optv [] = 
	{ 
		"34abcf:g:hklLmopPrstw:x", 
		oPUTOPTV_S_FILTER, 
		"format C/C++ source code with preamble annotations", 
		"3\tindent is 3 blanks", 
		"4\tindent is 4 blanks", 
		"a\tatheros standard format", 
		"b\tinsert preamble comment block", 
		"c\tconvert C++ comments to standard C comments", 
		"f s\tuse profile (s) [" LITERAL (PROFILE_NAME) "]", 
		"g s\tuse profile section (s) [" LITERAL (SECTION_NAME) "]", 
		"h\tinsert header include guard", 
		"k\tmake comments permanent", 
		"l\tupdate " LITERAL (oCOMMENT_S_LICENSE) " comment [" LITERAL (oCOMMENT_C_LICENSE) "]", 
		"L\tupdate " LITERAL (oCOMMENT_S_SPECIAL) " comment [" LITERAL (oCOMMENT_C_SPECIAL) "]", 
		"m\tconvert single-line comments to multi-line comments", 
		"o\toutput profile to stdout", 
		"p\tupdate " LITERAL (oCOMMENT_S_PACKAGE) " comment [" LITERAL (oCOMMENT_C_PACKAGE) "]", 
		"P\tupdate " LITERAL (oCOMMENT_S_PREFACE) " comment [" LITERAL (oCOMMENT_C_PREFACE) "]", 
		"r\tupdate " LITERAL (oCOMMENT_S_RELEASE) " comment [" LITERAL (oCOMMENT_C_RELEASE) "]", 
		"s\tinsert source include guard", 
		"t\tindent is 1 tab", 
		"w n\tbar width is (n)", 
		"x\terase markers", 
		(char const *) (0)
	}; 
	char const * profile = PROFILE_NAME; 
	char const * section = SECTION_NAME; 
	ogetoptv getopt; 
	ofileopen fileopen; 
	oprofile config; 
	octidy object; 
	signed (octidy::* method) (signed) = & octidy::charlie; 
	signed c; 
	while (~ (c = getopt.getoptv (argc, argv, optv))) 
	{ 
		switch (c) 
		{ 
		case '3': 
			object.offset ("   "); 
			break; 
		case '4': 
			object.offset ("    "); 
			break; 
		case 'a': 
			method = & octidy::atheros; 
			break; 
		case 'b': 
			object.setbits (oCOMMENT_B_COMMENT); 
			break; 
		case 'c': 
			object.setbits (oCOMMENT_B_DOUBLE); 
			break; 
		case 'f': 
			profile = getopt.args (); 
			break; 
		case 'g': 
			section = getopt.args (); 
			break; 
		case 'h': 
			object.label ("HEADER").state (1); 
			break; 
		case 'k': 
			object.setbits (oCOMMENT_B_PERMANENT); 
			break; 
		case 'l': 
			object.setbits (oCOMMENT_B_LICENSE); 
			break; 
		case 'L': 
			object.setbits (oCOMMENT_B_SPECIAL); 
			break; 
		case 'm': 
			object.setbits (oCOMMENT_B_TRIPLE); 
			break; 
		case 'o': 
			config.write (SECTION_NAME); 
			config.write (oCOMMENT_S_PREFACE, oCOMMENT_T_PREFACE); 
			config.write (oCOMMENT_S_PACKAGE, oCOMMENT_T_PACKAGE); 
			config.write (oCOMMENT_S_RELEASE, oCOMMENT_T_RELEASE); 
			config.write (oCOMMENT_S_LICENSE, oCOMMENT_T_LICENSE); 
			config.write (oCOMMENT_S_SPECIAL, oCOMMENT_T_SPECIAL); 
			exit (0); 
		case 'p': 
			object.setbits (oCOMMENT_B_PACKAGE); 
			break; 
		case 'P': 
			object.setbits (oCOMMENT_B_PREFACE); 
			break; 
		case 'r': 
			object.setbits (oCOMMENT_B_RELEASE); 
			break; 
		case 's': 
			object.label ("SOURCE").state (1); 
			break; 
		case 't': 
			object.offset ("\t"); 
			break; 
		case 'w': 
			object.width (atoi (getopt.args ())); 
			break; 
		case 'x': 
			object.setbits (oCOMMENT_B_DISCARD); 
			break; 
		default: 
			break; 
		} 
	} 
	object.preface (config.string (profile, section, oCOMMENT_S_PREFACE, oCOMMENT_T_PREFACE)); 
	object.package (config.string (profile, section, oCOMMENT_S_PACKAGE, oCOMMENT_T_PACKAGE)); 
	object.release (config.string (profile, section, oCOMMENT_S_RELEASE, oCOMMENT_T_RELEASE)); 
	object.license (config.string (profile, section, oCOMMENT_S_LICENSE, oCOMMENT_T_LICENSE)); 
	object.special (config.string (profile, section, oCOMMENT_S_SPECIAL, oCOMMENT_T_SPECIAL)); 
	if (! getopt.argc ()) 
	{ 
		(object.* method) (std::cin.get ()); 
	} 
	while (getopt.argc () && * getopt.argv ()) 
	{ 
		if (fileopen.openedit (* getopt.argv ())) 
		{ 
			object.filename (* getopt.argv ()); 
			(object.* method) (std::cin.get ()); 
			fileopen.close (); 
		} 
		getopt++; 
	} 
	exit (0); 
} 

